decode.go

```diff
@@ -397,17 +397,20 @@ func (d *decoder) callObsoleteUnmarshaler(n *Node, u obsoleteUnmarshaler) (good
 // unmarshalling was already done by UnmarshalYAML, and if so whether
 // its types unmarshalled appropriately.
 //
-// If n holds a null value, prepare returns before doing anything.
+// If n holds a null value, prepare will not initialize any values, but will
+// call unmarshallers, in order to allow resetting default values to null.
 func (d *decoder) prepare(n *Node, out reflect.Value) (newout reflect.Value, unmarshaled, good bool) {
-	if n.ShortTag() == nullTag || n.Kind == 0 && n.IsZero() {
-		return out, false, false
-	}
+	orig := out
 	again := true
 	for again {
 		again = false
 		if out.Kind() == reflect.Ptr {
 			if out.IsNil() {
-				out.Set(reflect.New(out.Type().Elem()))
+				// Only if the value we have is not null, to avoid setting null fields
+				// to their default value.
+				if n.ShortTag() != nullTag && n.Kind != 0 && !n.IsZero() {
+					out.Set(reflect.New(out.Type().Elem()))
+				}
 			}
 			out = out.Elem()
```

Comment:
This is only safe here because we're guaranteed to not be nil given the logic above. Once the above is conditional, this won't return a good out value. Am I missing something?

Reply:
No and given your other refactor this can get a bit more simple too. I added some more tests too.